### PR TITLE
Implemented filter for debug console output

### LINKED ISSUE #20838: vscode debug console needs to support error/log/info filters
image

Some sort of filter functionalaity like chrome devtools console or F12 console allowing one to filter errors/log/info would be nice. Being able to just see errors would drastically reduce signal to noise ratio when fixing unexpected js exceptions.

### LINKED ISSUE #93750: Debug Console NEEDS a filter
Issue Type: <b>Feature Request</b>

There is a lot of extensions that overwhelm the debug console (including C# and Flutter extensions). The Flutter extension team blames VSCode for not having the proper API's to allow filtering.

On Flutter it's even a bit more complicated, because the guy who build the extension is not the same guy who build the Flutter plugin that is not the same guy who wrote the original native plugin that is spamming the debug console.

So, WE NEEED A FILTER ON VSCODE (a simple "contains" and "starts with") will do. It is not that hard to implement =(

VS Code version: Code 1.43.2 (0ba0ca52957102ca3527cf479571617f0de6ed50, 2020-03-24T07:38:38.248Z)
OS version: Windows_NT x64 10.0.18362

### PR DESCRIPTION
**Pull request by arhelmus**
This PR fixes https://github.com/microsoft/vscode/issues/20838

After migration from Chrome DevTools to VS Code debugger, our users experienced major slowdown in productivity due to missed filtering in debug console. Users used to use filter to cut off irrelevant logs (as those pretty verbose) and with this PR I bringing this possibility in VS Code.

This feature is different from already implemented search, as instead of helping you to find what you searching for, it showing you only what you need.

Video: https://streamable.com/tvatmz

### CODE REVIEW

**Comment by isidorn** [comment]
@ArchDev thanks a lot for this great PR.
Since we are wrapping up our milestone and I was on vacation until today I am assigning this to August so we ideally merge it then and test it out.

In the meantime I just gave this a birds eye review.

1. Did you checkout how the markers view is doing the filtering [here](https://github.com/microsoft/vscode/blob/6f21ef0a18416916f86ef67d7218a95c79b68764/src/vs/workbench/contrib/markers/browser/markersTreeViewer.ts#L512). What needs to be done in the REPL is pretty much the same work unless I am mistaken
2. Please get latest from master so we test this out with latest Electron version which we pushed to master
3. Is it possible to write some unit tests
4. Is it possible to have code re-use between the markers view and what you introduced in the REPL, so we do not have everything duplicated

Next milestone I plan to do a more thorough review. 
Thanks a lot for your work!

fyi @joaomoreno and @sandy081 since they worked on the Markers view filter

**Comment by ArchDev** [comment]
Thanks for raising the priority for this @isidorn!

1. It was the plan initially, but to implement filtering by type (show infos/warnings/errors), this information should present as part of IReplElement. I didnt found easy way to get this information in, so decided to just remove this type of filtering at all.
2. 3. I will try to allocate some time this week to do this
4. It is definitely possible, I believe the input view itself can definetely by moved to common component. I don't feel confident in making this as I barely understand all the usecases of marker filter (I dropped quite a lot of code from it for repl), would love to get some help here.

**Comment by isidorn** [comment]
No problem, there is no priority, I was just on vacation for the past 3 weeks.

1. I did not mean filtering by type, but I think you answered my question, you started with what Markers pane is doing and tried to make it simpler
2 / 3 Great.
4. Let's look into this once we merge this PR, so we can do that later if we think it makes sense

Once you do 2/3 I plan to review it some time start of August. Thanks for working on this.

**Comment by arhelmus** [comment]
@isidorn rebased on master, ensured that it still works and compiles, added tests on filtering process in repl model

**Comment by isidorn** [comment]
@arhelmus great thanks. I plan to review and test it in detail in 10 days - since next week we are releasing our current stable build. I will provide more feedback then.

**Comment by isidorn** [comment]
Thanks again for this great PR. I have just tested this out and here's my feedback regarding the functionality:

1. Filter box background seems to not be themable, it should have the same background color as the problems filter. You can for example swithc to Monoaki theme
2. Even though the width is the same as the problems panel filter width I personaly think it is too wide. @sandy081 would you mind if we decrease the width of the filter box a bit?
3. Do not store filter state across vscode releoads. Reload should just clear filter. We can add this later if users ask for it

Regarding the code changes I will comment directly in the code.

File: src/vs/workbench/contrib/debug/browser/repl.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/repl.ts:458]: I do not understand why we update teh filterStat layout when in practice this does not seem to do anything ofr me.
Also filterState and filter should idealy not be attributes on the repl but just local variables passed to the tree.

File: src/vs/workbench/contrib/debug/browser/replFilter.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/replFilter.ts:1]: Is this a 1:1 copy from the Problems? If yes we should try to reuse code and have a shared filter class.
Though I can also look into this afterwards

fyi @sandy081

File: src/vs/workbench/contrib/debug/common/replModel.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/common/replModel.ts:186]: Is this really needed? Do the problems do the same?
I thought the tree will do the filtering even if we give it more data? Or I might be mistaken?

File: src/vs/workbench/contrib/debug/test/browser/repl.test.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/test/browser/repl.test.ts:194]: Good job with tests

File: src/vs/workbench/contrib/debug/browser/replFilterView.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/replFilterView.ts:158]: I believe this should just be in the previous `if` statment and we do not need the `handled` boolean. Feels like a copy paste error.

File: src/vs/workbench/contrib/debug/browser/replFilterView.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/replFilterView.ts:166]: This does not seem to work for me. Never does the filter grow, it is always large.

File: src/vs/workbench/contrib/debug/browser/replFilterView.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/replFilterView.ts:1]: As far as I understand nothing in this calss is REPL specific.
Is this a 1:1 copy from the Problems? If yes we should just have a shared class
fyi @sandy081

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
I have added comments inline. Mostly smaller things, looks good overall.
I would like to understand exactly how much fo the funcionality is shared between problems and debug console so ideally we would just share the same classes and not have code duplication.
Thanks a lot!

File: src/vs/workbench/contrib/debug/browser/media/replFilter.css
Comment (arhelmus) [reply src/vs/workbench/contrib/debug/browser/media/replFilter.css:1]: 100% same

File: src/vs/workbench/contrib/debug/common/replModel.ts
Comment (arhelmus) [reply src/vs/workbench/contrib/debug/common/replModel.ts:186]: unfortunately, this model is literally a model of real data source which has nothing to do with actual implementation (ReplDataSource). I didn't found better way to test this than to add support of primitive filtering to the model.

**Comment by arhelmus** [comment]
I will allocate some time this week to address your comments, thanks for review

**Comment by isidorn** [comment]
Fantastic. After discussing with my collegues we decided it would be best to introduce a new folder
`vs/workbench/contrib/treeFilter` under which we should put all the shared things between the Problems view and the DebugConsole. So the `.css`file and the `ParsedQuery` logic and everything which makes sense to you.

File: src/vs/workbench/contrib/debug/browser/repl.ts
Comment (arhelmus) [reply src/vs/workbench/contrib/debug/browser/repl.ts:129]: I was surprised with this comment as that's how filter stuff created in markers view, in constructor. I tried to move it to render body and problem is that repl.getActions is triggered before repl.renderBody as executed, and if filter state will not be created in constructor, we will need to handle undefined invariant in filter view

File: src/vs/workbench/contrib/debug/browser/repl.ts
Comment (arhelmus) [reply src/vs/workbench/contrib/debug/browser/repl.ts:458]: I think it is required for this state change, but I not sure how it is implemented for markers filter.

Screenshot 2020-08-16 at 12 50 00
Screenshot 2020-08-16 at 12 49 55

File: src/vs/workbench/contrib/debug/browser/replFilterView.ts
Comment (arhelmus) [reply src/vs/workbench/contrib/debug/browser/replFilterView.ts]: fixed

**Comment by arhelmus** [comment]
@isidorn I fixed 1 and 3 from your first comment, moved the filter view in separate component (I believe its only part which can be shared and its actually the part with all the copy-paste I did) and addressed your comments to the code.

Do you think now you can handle this PR to unify treeViewItem between markers and repl?

**Comment by isidorn** [comment]
@arhelmus thanks a lot.
I can try to take it over from here to merge it in and also use it in the markers view.

File: src/vs/workbench/contrib/debug/browser/media/replFilter.css
Comment (sandy081) [reply src/vs/workbench/contrib/debug/browser/media/replFilter.css]: works for me

**Comment by sandy081** [comment]
@isidorn Just back from vacation. Have not gone through all changes and I trust your decisions. Let me know if you need any input from me?

**Comment by isidorn** [comment]
@sandy081 ok. I will take care of it. Will ping you if something is needed. Thanks

**Comment by RobinLbt** [comment]
@isidorn is it included in the last release of VS Code ? I can't find filters in the debug console still

**Comment by isidorn** [comment]
It is in vscode insiders. Will be in next vscode stable in 10 days.

**Pull request merged**
